### ISSUE TITLE
UINAvigationController -> setViewControllers causes a crash

### ISSUE THREAD

**Issue by yaroslav-v** [ReactiveX/RxSwift#1271]
**Short description of the issue**:

Call of the method UINAvigationController -> setViewControllers:animated: causes a crash of the app. It's happening only on iOS 10.3.2 and when I'm building the app in release mode.

The issue appears on iOS 10.3.2 and in release mode only. I’ve checked this on iPhone with 10.3.2 and release build fails but debug works OK. Debug and release builds work fine on all of the previous versions of iOS.

There is no debug log or any other message which I can provide you because the app just disappears from the screen with no notification at all.

**Cause**:

I've found that this behavior was introduced after update of RxCocoa from 3.3.1 to 3.4.0. It happens because of the following change in DelegateProxyType.swift : extension ObservableType : func subscribeProxyDataSource:

```swift
	 return Disposables.create { [weak object] in
		 subscription.dispose()
-		unregisterDelegate.dispose()
		 object?.layoutIfNeeded()
+		unregisterDelegate.dispose()
	 }
```

**Expected outcome**:

User should see new view controller on the screen.

**What actually happens**:

The app just disappears from the screen with no notification at all.

**Just short example of code to see how it called**:

AppDelegate.swift
```swift
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
    ...
    window = UIWindow(frame: UIScreen.main.bounds)
    ....
    let navigationController = UINavigationController(rootViewController: viewController)
    window.rootViewController = navigationController
    window.makeKeyAndVisible()
}
```

ViewController.swift
```swift
override func viewDidLoad() {
    super.viewDidLoad()
    ...
	continueButton.rx.tap
		.asDriver()
		.drive(onNext: { [unowned self] value in
			let nextController = FinalBuilder.viewController()
			navigationController?.setViewControllers([nextController], animated: true)
		})
		.addDisposableTo(disposeBag)
    ...
}
```

**RxSwift/RxCocoa/RxBlocking/RxTest version/commit**

  3.4.0 and above

**Platform/Environment**

  - [ x] iOS
  - [ ] macOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] playgrounds

**How easy is to reproduce? (chances of successful reproduce after running the self contained code)**

  - [ x] easy, 100% repro
  - [ ] sometimes, 10%-100%
  - [ ] hard, 2% - 10%
  - [ ] extremely hard, %0 - 2%

**Xcode version**:

```
 8.3.2
```

**Installation method**:
  - [ x] CocoaPods
  - [ ] Carthage
  - [ ] Git submodules

**I have multiple versions of Xcode installed**:
  (so we can know if this is a potential cause of your issue)
  - [ ] yes (which ones)
  - [ x] no

**Level of RxSwift knowledge**:
  (this is so we can understand your level of knowledge
    and formulate the response in an appropriate manner)
  - [ ] just starting
  - [ x] I have a small code base
  - [ ] I have a significant code base

**Comment by kzaher** [comment]
Hi @yaroslav-v ,

I'm trying to repro this. Unfortunately since I've upgraded my device to 10.3.2 I can't see the device any more in either iTunes or Xcode :(, and there is no available update for either application.

What Xcode/iTunes version do you have?

**Comment by kzaher** [comment]
Oh, good old macOS restart fixed the hidden device issue.

But now I still can't repro this. I've tried using `master` branch version and 

```
let window = UIWindow(frame: UIScreen.main.bounds)
        let navigationController = UINavigationController(rootViewController: UIViewController())
        window.rootViewController = navigationController
        window.makeKeyAndVisible()

        _ = navigationController.rx.didShow.subscribe(onNext: { x in
            print(x)
        })
        
        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(1000), execute: {
            let nextController = UIViewController()
            navigationController.setViewControllers([nextController], animated: true)
        })
```

... but this doesn't crash the app. I get this

```
(viewController: <UIViewController: 0x105e15dc0>, animated: false)
(viewController: <UIViewController: 0x105e15dc0>, animated: false)
(viewController: <UIViewController: 0x105e2abf0>, animated: true)
```

Do you have some self contained code sample I can use to repro this?

**Comment by yaroslav-v** [comment]
hi, I'm using Xcode 8.3.2 and iTunes 12.6.0.100. I'll try to prepare self contained code sample and let you know :).

**Comment by kzaher** [comment]
@yaroslav-v any news on this? We are planning to ship next release soon, so if we can fix this I would like to include this issues.

**Comment by yaroslav-v** [comment]
hi, I've tried to distill spoiled code from my project but with no luck. But I see that it maybe be related to table view in one of view controllers, which were stored in navigation controller when I tried to use setViewControllers.

**Comment by kzaher** [comment]
Ok, if you can isolate the problem and if it's caused by RxSwift feel free to reopen this issue.

According to what you've said it seems to me that RxSwift only triggered layout call that for some reason caused the crash, but there aren't any obvious bugs in the library, so I'm going to close this for now.

**Comment by pitiphong-p** [comment]
I think I have the same issue as this one. I have a table view binding with `rx.items(cellIdentifier:cellType:)` method and have a `rx.itemSelected` binding with unwind segue action in its subscribe block.
What I found is that sometimes it crashes in the items block for using the dealloced `self` (with `unowned` capturing). The condition is that the list must be longer than the table view itself.
I also found that this caused by the table view try to `layoutSubviews` in the dispose block and the index send to the `items` block is the index of the upcoming cell so I think this issue may be related to the `Cell prefetching` feature introduced in iOS 10
